docs(deepagents): `RubricMiddleware` (#4183)
langchain-ai/docs: src/oss/deepagents/rubric-middleware.mdx

> ---

Reviewer: i would rename this page to just `rubric.mdx`

title: Rubrics
description: Self-evaluating agents that iterate against a rubric until done
---

> Some agent tasks have a clear definition of *done* that the model alone cannot reliably hit on the first try: a haiku in the right syllable pattern, a refactor with all tests passing, a report that hits every required section. `RubricMiddleware` lets you declare *what done looks like* as a rubric and have the agent **self-evaluate and iterate** until the rubric is satisfied (or a configured maximum iteration cap is hit).

Reviewer: ```suggestion
Some agent tasks have a clear definition of "done" that the model alone cannot reliably hit on the first try: a haiku in the right syllable pattern, a refactor with all tests passing, a report that hits every required section. `RubricMiddleware` lets you declare *what done looks like* as a rubric and have the agent **self-evaluate and iterate** until the rubric is satisfied (or a configured maximum iteration cap is hit).
```

> It mounts two hooks. `before_agent` detects a new rubric and resets the iteration counter. `after_agent` runs a separate **grader sub-agent** that reviews the transcript against the rubric. If the grader returns `needs_revision`, the middleware injects the feedback as a `HumanMessage` and jumps back to the model. The loop terminates on `satisfied`, `max_iterations_reached`, `failed`, or a real interrupt.

Reviewer: I'm not sure how useful it is to go into implementation details here by naming hooks -- but could be convinced otherwise

```mermaid
graph LR

>     Start[User invokes<br/>with rubric] --> Agent[DeepAgent]

Reviewer: ```suggestion
    Start[User invokes<br/>with rubric] --> Agent[Deep agent]
```

    Agent --> Grader{Grader<br/>verdict}

    Grader --> |satisfied| Done[Finish execution]
    Grader --> |failed| Done
    Grader --> |needs_revision| Cap{Iterations < <br/> max_iterations?}

>     Cap --> |yes| Inject[Re-prompt DeepAgent with per-criterion feedback]

Reviewer: ```suggestion
    Cap --> |yes| Inject[Re-prompt deep agent with per-criterion feedback]
```

    Cap --> |no| Done

    Inject --> Agent

    classDef trigger fill:#F6FFDB,stroke:#6E8900,stroke-width:2px,color:#2E3900
    classDef process fill:#E5F4FF,stroke:#006DDD,stroke-width:2px,color:#030710
    classDef decision fill:#FDF3FF,stroke:#7E65AE,stroke-width:2px,color:#504B5F
    classDef alert fill:#F8E8E6,stroke:#B27D75,stroke-width:2px,color:#634643

    class Start trigger
    class Agent,Inject process
    class Grader,Cap decision
    class Done,MaxOut alert
```

> When no `rubric` is supplied on input state, both hooks short-circuit and the middleware does not run, so it is safe to include unconditionally in a `create_deep_agent` stack.

Reviewer: ```suggestion
```

<Note>

> `RubricMiddleware` is in **beta**. The API may change in the future.

Reviewer: would consider x-ref'ing to https://docs.langchain.com/oss/python/versioning

</Note>

## Rubric verdicts

Every grader pass produces one of four verdicts:

| Status | Meaning | Loops back? |
| --- | --- | --- |
| `satisfied` | Every criterion in the rubric passes. | No |
| `needs_revision` | At least one criterion fails; grader feedback is injected and the agent runs again. | Yes |
| `max_iterations_reached` | Grader still wants revisions, but `max_iterations` has been hit. | No |
| `failed` | The grader could not produce a verdict; either it judged the rubric un-gradable, or the grader call raised an exception. | No |

> To exit the agent mid-run, press Ctrl+C. `KeyboardInterrupt` and `asyncio.CancelledError` are not caught, so keyboard interrupts and async cancellation propagate out of `agent.invoke()` as expected.

Reviewer: ```suggestion
```

> 

Reviewer 2: Should this be with `failed`? Or shoudl this be here at all? IT seems odd in a rubric verdicts section

## Configuration

> `RubricMiddleware` takes one required argument and four optional ones:

Reviewer 2: Best not to specifiy as these tend to change over time and updating this text is easily forgotten
```suggestion
```

| Argument | Required | Default | Description |
| --- | --- | --- | --- |

> | `model` | Yes | `None` | Chat model used by the grader sub-agent. Accepts a `"provider:model-id"` string or a `BaseChatModel` instance. |

Reviewer 2: interesting - why doesn't it use the agent model as the default? doesn't need to be in the docs but I'm curious

Author: Possibly, but we decided to leave the grader model choice to the user - the deep agent + grader loop can get long and expensive, and making it explicit gives users a clearer picture of the possible token costs/usage. That said, we could revisit this in a future version by defaulting the grader to the same model as the deep agent if none is specified.

| `system_prompt` | No | Built-in grader prompt | Custom grading instructions. Falls back to a default system prompt that teaches the grader the verdict format and what tools it has at its disposal. |
| `tools` | No | `None` | Tools the grader may call to gather evidence (run tests, count tokens, read files) before producing a verdict. With none, the grader reasons from the transcript alone. |
| `max_iterations` | No | `3` | Hard cap on grader iterations per rubric attempt. Maximum input value is 20. When the cap is reached without a `satisfied` verdict, the agent terminates with status `max_iterations_reached`. |
| `on_evaluation` | No | `None` | Optional callback invoked with each `RubricEvaluation` after grading. Useful for evals, observability, or persisting iteration history. |

> If no rubric is passed on invocation state, the middleware no-ops — the agent runs once and returns as normal.

Reviewer 2: This is the configuration section  - so shouldn't have invocation info
```suggestion
```

> ## Example: Grading a Haiku with a Custom Tool and Prompt:

Reviewer: ```suggestion
## Example: grading a haiku with a custom tool and prompt
```

> The example below builds a deep agent that writes a haiku about the ocean. The rubric asks for a 5/7/5 syllable pattern, imagery that is specifically oceanic (not generic "water" or "beach"), and concrete sensory detail.

Reviewer 2: ```suggestion
The following example builds a deep agent that writes a haiku about the ocean. The rubric asks for a 5/7/5 syllable pattern, imagery that is specifically oceanic (not generic "water" or "beach"), and concrete sensory detail.
```

> ```python

Reviewer: ```suggestion
```python expand
```

            system_prompt=HAIKU_GRADER_PROMPT,
            tools=[count_syllables],
            max_iterations=5,
            on_evaluation=lambda ev: print(ev["result"], ev["explanation"]),
        ),
    ],
    checkpointer=InMemorySaver(),
)

config = {"configurable": {"thread_id": "haiku-session"}}
result = agent.invoke(
    {
        "messages": [HumanMessage("Write a haiku about the ocean.")],
        "rubric": (
            "- 5/7/5 syllable pattern\n"
            "- imagery is specifically oceanic (waves, tide, salt, kelp, etc.).\n"
            "- uses concrete sensory detail (sight, sound, smell, touch)"
        ),
    },
    config=config,
)

print(result["messages"][-1].content)   # the final haiku
```

The `rubric` field on input state is the trigger. Each grader iteration is delivered to `on_evaluation` as a `RubricEvaluation` dictionary containing the verdict, explanation, and per-criterion gaps, which is useful for evals, dashboards, or persisting an audit trail.

## Rubric persistence

A single `agent.invoke()` call runs the rubric loop to completion and returns with a terminal verdict: `satisfied`, `failed`, or `max_iterations_reached`.

> Whether the rubric carries over to a follow-up invocation depends on three things: is a checkpointer attached, was a `thread_id` supplied, and is it the same one? In each scenario below, the second `invoke` does not pass a `rubric`:

Reviewer 2: This is confusing because I don't know in which case it carries over. I think this is clearer
```suggestion
Rubrics carry over to follow-up invocations only if a checkpointer is attached and the same `thread_id` is passed alongside the invocation.
The second `invoke` reuses the same rubric unless you pass a new `rubric:
```
Can the second invoke pass the same rubric? will that mess things up? If so this needs to be amended.

**No checkpointer.** State is not persisted, so the second call has no rubric and the middleware no-ops.

```python
agent = create_deep_agent(model=..., middleware=[RubricMiddleware(...)])

agent.invoke({
    "messages": [HumanMessage("Write a launch tweet for the v2.4 SDK release.")],
    "rubric": (
        "- Under 280 characters\n"
        "- Mentions the performance gains\n"
        "- No emojis"
    ),
})
agent.invoke({
    "messages": [HumanMessage("Draft the release-notes blog post for v2.4.")],
})  # no grading
```

> **Checkpointer + same `thread_id`.** LangGraph rehydrates the thread's state, including the rubric from the prior invocation. The grader evaluates the new response against it. Pass a new `rubric` in the input dictionary to replace it.

Reviewer 2: I think none of this is needed here
```suggestion
```

```python
agent = create_deep_agent(..., checkpointer=InMemorySaver())
cfg = {"configurable": {"thread_id": "T1"}}

agent.invoke({
    "messages": [HumanMessage("Write a launch tweet for the v2.4 SDK release.")],
    "rubric": (
        "- Under 280 characters\n"
        "- Mentions the performance gains\n"
        "- No emojis"
    ),
}, config=cfg)
agent.invoke({
    "messages": [HumanMessage("Draft the release-notes blog post for v2.4.")],
}, config=cfg)  # graded against the tweet rubric
```

**Checkpointer + different `thread_id`.** A new thread starts empty, so the second call has no rubric and the middleware no-ops.

```python
agent = create_deep_agent(..., checkpointer=InMemorySaver())

agent.invoke({
    "messages": [HumanMessage("Write a launch tweet for the v2.4 SDK release.")],
    "rubric": (
        "- Under 280 characters\n"
        "- Mentions the performance gains\n"
        "- No emojis"
    ),
}, config={"configurable": {"thread_id": "T1"}})
agent.invoke({
    "messages": [HumanMessage("Draft the release-notes blog post for v2.4.")],
}, config={"configurable": {"thread_id": "T2"}})  # no grading

> ```

Reviewer 2: I don't think this is helpful as an example here
```suggestion
```
